tls/s2n_async_pkey.c

```diff
@@ -409,3 +422,140 @@ S2N_RESULT s2n_async_pkey_sign_free(struct s2n_async_pkey_op *op)
 
     return S2N_RESULT_OK;
 }
+
+int s2n_async_pkey_op_get_op_type(struct s2n_async_pkey_op *op, s2n_async_pkey_op_type *type)
+{
+    POSIX_ENSURE_REF(op);
+    POSIX_ENSURE_REF(type);
+
+    switch (op->type) {
+        case S2N_ASYNC_DECRYPT:
+            *type = S2N_ASYNC_DECRYPT;
+            break;
+        case S2N_ASYNC_SIGN:
+            *type = S2N_ASYNC_SIGN;
+            break;
+        default:
+            POSIX_BAIL(S2N_ERR_ASYNC_OP_UNINITIALIZED);
+            break;
+    }
+
+    return S2N_SUCCESS;
+}
+
+int s2n_async_pkey_op_get_input_size(struct s2n_async_pkey_op *op, uint32_t *data_len)
+{
+    POSIX_ENSURE_REF(op);
+    POSIX_ENSURE_REF(data_len);
+
+    const struct s2n_async_pkey_op_actions *actions = NULL;
+    POSIX_GUARD_RESULT(s2n_async_get_actions(op->type, &actions));
+    POSIX_ENSURE_REF(actions);
+
+    POSIX_GUARD_RESULT(actions->get_input_size(op, data_len));
+
+    return S2N_SUCCESS;
+}
+
+static S2N_RESULT s2n_async_pkey_get_input_size_decrypt(struct s2n_async_pkey_op *op, uint32_t *data_len)
+{
+    struct s2n_async_pkey_decrypt_data *decrypt = &op->op.decrypt;
+    struct s2n_blob *in = &decrypt->encrypted;
+
+    *data_len = in->size;
+    return S2N_RESULT_OK;
+}
+
+static S2N_RESULT s2n_async_pkey_get_input_size_sign(struct s2n_async_pkey_op *op, uint32_t *data_len)
+{
+    struct s2n_async_pkey_sign_data *sign = &op->op.sign;
+    struct s2n_hash_state *digest = &sign->digest;
+
+    uint8_t digest_length = 0;
+    RESULT_GUARD_POSIX(s2n_hash_digest_size(digest->alg, &digest_length));
+
+    *data_len = digest_length;
+
+    return S2N_RESULT_OK;
+}
+
+int s2n_async_pkey_op_get_input(struct s2n_async_pkey_op *op, uint8_t *data, uint32_t data_len)
+{
+    POSIX_ENSURE_REF(op);
+    POSIX_ENSURE_REF(data);
+
+    const struct s2n_async_pkey_op_actions *actions = NULL;
+    POSIX_GUARD_RESULT(s2n_async_get_actions(op->type, &actions));
+    POSIX_ENSURE_REF(actions);
+
+    POSIX_GUARD_RESULT(actions->get_input(op, data, data_len));
+
+    return S2N_SUCCESS;
+}
+
+static S2N_RESULT s2n_async_pkey_get_input_decrypt(struct s2n_async_pkey_op *op, uint8_t *data, uint32_t data_len)
+{
+    struct s2n_async_pkey_decrypt_data *decrypt = &op->op.decrypt;
+    struct s2n_blob *in = &decrypt->encrypted;
+
+    RESULT_ENSURE_LTE(in->size, data_len);
+
+    RESULT_CHECKED_MEMCPY(data, in->data, in->size);
+
+    return S2N_RESULT_OK;
+}
+
+static S2N_RESULT s2n_async_pkey_get_input_sign(struct s2n_async_pkey_op *op, uint8_t *data, uint32_t data_len)
+{
```

Comment:
Can this method be called twice for the same op?

Reply:
I think that would be best. I will update the code to copy the digest instead of consuming it

Reply:
Can you make sure you add unit tests for calling each of these twice? Between this and the alloc/realloc, seems like it's an easy place to make a mistake.

Reply:
Yes will add this